p4-16/spec/P4-16-draft-spec.mdk

```diff
@@ -2647,196 +2645,191 @@ X.v1  // reference to v1
+Note that bitwise operations on signed integers are well-defined, since the
 representation is mandated to be 2's complement.
 
 The ```int<W>``` datatype supports the following operations; all
 binary operations require both operands to have the exact same
 type. The result always has the same width as the left operand.
 
-- Negation, denoted by unary ```-```
-- Unary plus, denoted by ```+```. Behaves as a no-op.
-- Addition, denoted by ```+```
-- Subtraction, denoted by ```-```
-- Comparison for equality and inequality ```==,!=``` producing a
-  boolean result
-- Numeric comparisons ```<,<=,>,>=``` with a boolean result
-- Multiplication ```*```. Result has the same width as the
-  operands. P4 targets may impose additional restrictions (e.g., may
-  only allow multiplications with powers of two).
+- Negation, denoted by unary ```-```.
+- Unary plus, denoted by ```+```. This operation behaves like a no-op.
+- Addition, denoted by ```+```.
+- Subtraction, denoted by ```-```.
+- Comparison for equality and inequality, denoted ```==``` and ```!=``` respectively. These operations produce a
+  boolean result.
+- Numeric comparisons, denoted by ```<,<=,>,``` and ```>=```. These operations produce a boolean result.
+- Multiplication, denoted by ```*```. Result has the same width as the
+  operands. P4 architectures may impose additional restrictions---e.g., they may
+  only allow multiplication by a power of two.
 - Arithmetic shift left and right denoted by ```<<``` and ```>>```.
-  Left operand is signed, right operand must be either an
-  unsigned number of type ```bit<S>``` or a non-negative constant
-  integer. The result has the same type as the left operand. Shifts
-  with an amount greater or equal to the width of the input are
-  allowed.
+  The left operand is signed and the right operand must be either an
+  unsigned number of type ```bit<S>``` or a non-negative
+  integer literal. The result has the same type as the left operand. Shifting by an amount greater than
+  the width of the input produces a result where all bits are zero.
 
 ### A note about shifts
 
 Shifts (on signed and unsigned values) deserve a special discussion
 for the following reasons:
 
-- As in C, right shift behaves differently for signed and unsigned
+- Right shift behaves differently for signed and unsigned
   values: right shift for signed values is an arithmetic shift.
 - Shifting with a negative amount does not have a clear semantics:
-  while in C the result is undefined, in P4 the type system makes it
+   the P4 type system makes it
   illegal to shift with a negative amount.
-- In C, shifting with an amount larger or equal to the number of bits
-  has an undefined result (unlike the P4 definition).
-- Finally, shifting may require doing work which is exponential in the
-  number of bits of the right-hand-side operand.
+- Unlike C, shifting by an amount larger or equal to the number of bits
+  has a well-defined result (unlike C).
```

Comment:
Thanks. In general, I think we should avoid comparing P4 to C/C++/Java so much. There are sections in the spec where it seems to be every other paragraph and I don't think it helps promote understanding. I've been removing these as I go when it seems sensible.